Use a connection pool for communication with Solr

mwoodiupui opened the pull request:
## References
* Fixes #7986

## Description
Instead of having every use of Solr create and destroy a fresh connection, use a common pool of open connections.

## Instructions for Reviewers
List of changes in this PR:
* New HttpConnectionPoolService to get connections from a pool.
* Patch all uses of Solr to use pooled connections.

**Include guidance for how to test or review your PR.** This may include: steps to reproduce a bug, screenshots or description of a new feature, or reasons behind specific changes. 

## Checklist
- [x] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & integration tests). Exceptions may be made if previously agreed upon.
- [x] My PR passes Checkstyle validation based on the [Code Style Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Style+Guide).
- [x] My PR includes Javadoc for _all new (or modified) public methods and classes_. It also includes Javadoc for large or complex private methods.
- [x] My PR passes all tests and includes new/updated Unit or Integration Tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [x] If my PR includes new, third-party dependencies (in any `pom.xml`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [x] If my PR modifies the REST API, I've linked to the REST Contract page (or open PR) related to this change.

mwoodiupui committed:
First attempt at a single pool of HTTP connections for Solr.

mwoodiupui committed:
Define pool configuration properties

mwoodiupui committed:
Make pool service injectable.

mwoodiupui committed:
Initialize correctly.  More generous pool defaults.

mwoodiupui committed:
Make connection maximum lifetime configurable. #7986

mwoodiupui commented:
The pool size settings and timeouts are rather arbitrary.

mwoodiupui committed:
Add unit test.  #7986

tdonohue commented on dspace-api/src/test/java/org/dspace/service/impl/HttpConnectionPoolServiceTest.java:
```diff
@@ -0,0 +1,96 @@
+     * @throws java.net.URISyntaxException when an invalid URI is constructed.
+     */
+    @Test
+    public void testGetClient()
```
This is a good basic "smoke test". I wonder if it'd be possible to add a few other tests...for example.

* Set `solr.client.maxTotalConnections` to 1 and verify two calls to getClient() end up with the same connection.
* Set `solr.client.keepAlive` to a small value, and verify the connection auto-closes after that time.
* Set `solr.client.timeToLive` to a small value, and verify connection ejected from pool

These are just some ideas of ways we might add a few more additional tests here. But, if any of these seem really hard to easily verify in code, I'm OK with moving forward with this PR with minimal tests.

tdonohue replied:
@mwoodiupui : Previously I had asked if it'd be possible to add a few more tests here to verify each of the basic `solr.client.*` configs are working as expected?  See https://github.com/DSpace/DSpace/pull/7987/files#r766220588  As noted there, if these tests seem too difficult to build, let me know.  We can move this PR forward without these tests, but I wanted to ask whether these tests would be possible to add.

mwoodiupui replied:
There may not be a good way to do this.  The ClientBuilder uses its configured ConnectionManager to get a connection.  There seems no good way to get at the actual connection to see if it is an identical reference to another client's connection.  The actual pool is hidden inside ConnectionManager and well insulated from the outside world.

tdonohue replied:
@mwoodiupui : Ok, we can leave it as is then.  I was just trying to see what was possible & I trust your judgment here. I'll add this to my list to re-review since it looks like you've updated it today.

tdonohue reviewed (commented):
@mwoodiupui : Looks good overall!  I've tested this today (both with a full reindex & searching/browsing from the UI). I'm noticing no change in behavior (which is what we'd want to see).  I don't have a way to easily test the initial bug in #7986 no longer occurs, but this code _seems like it should fix that_, and it's not impacting normal behavior.

That said, I have a few inline notes below on some changes we may want to make.  Nothing seems major though.

mwoodiupui commented on dspace-api/pom.xml:
```diff
@@ -934,6 +932,79 @@
 
+    <dependencyManagement>
+        <dependencies>
+          <!-- for mockserver -->
```
Yes, mockserver needs newer versions of a number of its dependencies that we share.

mwoodiupui commented on dspace-api/src/main/java/org/dspace/service/impl/HttpConnectionPoolService.java:
```diff
@@ -0,0 +1,168 @@
+    @PostConstruct
+    protected void init() {
+        connManager = new PoolingHttpClientConnectionManager(
+                configurationService.getIntProperty("solrClient.timeToLive", DEFAULT_TTL),
```
Right you are:  they should be `solr.client.*`.

mwoodiupui committed:
Clarify the need to manage dependencies for mockserver.  #7986

mwoodiupui committed:
Generalize HttpConnectionPoolService for multiple instances with separate configuration.

This removes Solr-specific configuration property names from the otherwise
general service implementation by adding a bean property to specify a
configuration property prefix.

mwoodiupui committed:
Autowire with 'id' not <qualifier>:  @Named doesn't find qualifiers.  #7986

mwoodiupui commented on dspace/config/dspace.cfg:
```diff
@@ -45,6 +45,20 @@ default.language = en_US
 # Since DSpace 7, SOLR must be installed as a stand-alone service
 solr.server = http://localhost:8983/solr
 
+# Solr connection pool.
```
In fact the ConnectionKeepAliveStrategy is instantiated when the enclosing bean is instantiated, so solr.client.keepAlive is also effectively fixed at startup time.  So noted.

tdonohue replied:
@mwoodiupui  Could we add comments to these dspace.cfg settings to specify that they require a servlet container reboot?  You had said "So noted", but I don't see the comments added...maybe you just forgot to push them up?

mwoodiupui replied:
I *thought* I had documented this, but I cannot find it anywhere.  Done.

mwoodiupui commented on dspace-api/src/main/java/org/dspace/service/impl/HttpConnectionPoolService.java:
```diff
@@ -0,0 +1,168 @@
+ */
+@Named
+@Singleton
+public class HttpConnectionPoolService {
```
Since the only Solr-specific thing is the names of the configuration properties, I've made the leading part of those names a constructor argument.  We can now have a pool for Solr configured by `solr.client.*`, a pool for X configured by `X.client` and so forth.  Code that wants a pool can have the one it needs injected by `id`.

mwoodiupui committed:
Merge branch 'main' into 7986

tdonohue commented on dspace-api/src/test/java/org/dspace/service/impl/HttpConnectionPoolServiceTest.java:
```diff
@@ -0,0 +1,96 @@
+        System.out.println("getClient");
+
+        configurationService.setProperty("solrClient.maxTotalConnections", 2);
+        configurationService.setProperty("solrClient.maxPerRoute", 2);
```
@mwoodiupui : It looks like this Test still is using the wrong config names. These should start with `solr.client.*`

tdonohue reviewed (commented):
@mwoodiupui : I gave this another review today...it looks like you did some cleanup based on my feedback (thanks!), but you overlooked a few other points of feedback.  See inline comments below.

mwoodiupui committed:
Correct the names of configuration properties set by tests.  #7986

mwoodiupui committed:
More documentation.  Note that DSpace restart is required for configuration changes.

tdonohue reviewed (approved):
👍 Looks good to me now, @mwoodiupui !  I've re-reviewed and the code/configurations make sense.  I've also tested this PR and it appears to be working well...and I'm noticing no change in behavior with normal indexing/re-indexing processes (which is what I'd expect).

Thanks! I"ll go ahead and merge this as the team previously agreed that it just needed one approval.

On your end, I think the only thing we need to do now is add these new configs to our official docs at https://wiki.lyrasis.org/display/DSDOC7x/Configuration+Reference

Thanks again!

mwoodiupui commented:
Documentation updated with a table of the new pool configuration properties.
